portal/models/flaskdanceprovider.py

```diff
@@ -11,21 +11,134 @@ class FlaskDanceProvider:
+        return self.parse_json(user_json)
+
+    def parse_json(self, user_json):
+        """parses the user's json and returns it in a standard format
+
+        Providers encode user information in json. This function parses
+        the json and stores values in an instance of FlaskProviderUserInfo
+
+        :param user_json: info about the user encoded in json
+
+        :return instance of FlaskProviderUserInfo with the user's info
+        """
+
+        def get_value_from_json(standard_key, required=True):
+            """gets a value from the provider json
+
+            Each provider returns json with different property values.
+            For example, Facebook returns json that maps a user's first name to
+            'first_name' while Google maps first names to 'given_name'.
+            In addition, a value could be nested. For example, Facebook's
+            profile picture is nested in json['picture']['data']['url'].
+            self.standard_key_to_provider_key_map links standard keys to
+            provider specific keys, including those that are nested, which
+            allows our parsing code to stay as generic as possible.
+
+            :param standard:key: the standard key
+            :param required: is this property required?
+            :return value from the user's json or None
+            """
+
+            # Ge the key used by the provider
+            user_json_key = \
+                self.standard_key_to_provider_key_map[standard_key]
+
+            # The key could have multiple nested parts
+            # e.g. 'picture.data.url'
+            # Which means we'll need to get each part
+            # individually
+            parts = user_json_key.split('.')
+
+            # Loop over each key to get the value
+            # from the user's json
+            value = user_json
+            for key in parts:
+                # Certain properties can be undefined
+                # Handle these cases gracefully
+                if not required and key not in value:
+                    return None
+
+                value = value[key]
+
+            return value
+
+        # Attempt to parse the user's json
+        try:
+            user_info = FlaskProviderUserInfo()
+            user_info.id = get_value_from_json('id')
+            user_info.first_name = get_value_from_json('first_name')
+            user_info.last_name = get_value_from_json('last_name')
+            user_info.email = get_value_from_json('email')
+            user_info.image_url = get_value_from_json('image_url')
+
+            # These properties may not be defined
+            user_info.gender = get_value_from_json(
+                'gender',
+                required=False
+            )
+            user_info.birthdate = get_value_from_json(
+                'birthdate',
+                required=False
+            )
+
+            return user_info
+        except KeyError as err:
+            current_app.logger.debug(
```

Comment:
If we're missing a required value, I would probably consider that an error, ie `current_app.logger.error`

Reply:
Agreed